JavaDoc and normal doc changes (quite a few of them TODO comments)

matshenricson opened the pull request.

RCasatta commented:
ack, just giving @lvaccaro some time to have a look

lvaccaro commented:
I agree on fixing the documentation. I don't like so much insert todos, in particular, if they are related to each others; better know them and fix soon, instead of adding and removing the TODO comment later, for example:
- in exceptions: the message of TODO could be put inside the exception
- in crypto operations: we can think later about how to raise exceptions and refactoring

lvaccaro commented on src/main/java/com/eternitywall/ots/Timestamp.java:
```diff
@@ -127,13 +118,14 @@ public void serialize(StreamSerializationContext ctx) {
                 sortedAttestations.get(sortedAttestations.size() - 1).serialize(ctx);
             }
-        } else if (!this.ops.isEmpty()) {
+        } else if (!this.ops.isEmpty()) {      // TODO: Isn't this always true? Fix later.
```
a leaf of the markle three, with attestations, has always 0 ops

matshenricson replied:
I was thinking strictly controll-flow here. My understanding of the underlying data structures is shaky.
But the code is like this:
`if (x) {
   // Do things
} else if (!x) {
}`
It is always the case that it is true, UNLESS "// Do things" changes x in some way. I did a quick look at the code and came up with my comment.

So, what do we do now? I promised this PR to be ONLY comments, which is why I did no code changes. Should I remove the comment, or change the code?

matshenricson replied:
OK, I removed this TODO, I think I will remember this issue later.

lvaccaro replied:
thanks for the explanation 👍

lvaccaro commented on src/main/java/com/eternitywall/ots/Timestamp.java:
```diff
@@ -242,7 +234,7 @@ public TimeAttestation shrink() throws Exception {
             Map.Entry<Op, Timestamp> entry = it.next();
             Timestamp timestamp = entry.getValue();
-            Op op = entry.getKey();
+            Op op = entry.getKey();    // TODO: Never used...
```
you can just remove it

matshenricson replied:
But this PR is ONLY comments and JavaDoc. I wanted to make this PR as simple as possible for you to accept. I will remove it in the next next PR, which you haven't seen yet. I wrote that code two days ago but haven't submitted it as a PR yet.

matshenricson replied:
I removed this TODO also, will remember to fix this later.

lvaccaro commented:
Running javadoc with
```
mvn javadoc:javadoc
```
For most of them, you have just made some changing in this PR in order to add comments (on functions there are not documented before) and make them better (cleanup line, fix typo, etc.. ) in lots methods :-)  but javadoc fails (as: coalesce, arraysCopy, arrayReverse, compare). Could you fix them?

matshenricson commented:
Do you want me to fix the JavaDoc issues before you accept this PR?

lvaccaro commented:
Yes, if it is possible.. I'd like not break the javadoc command.
@RCasatta we should insert javadoc also in travis CI

matshenricson commented:
I have fixed three issues now, but wasn't able to break JavaDoc myself. What did it report?

RCasatta commented:
there aren't issue on javadoc now, javadoc is not on travis CI now but it's mandatory for publishing the lib on maven

RCasatta commented:
ready to merge for me, @matshenricson do you want to squash your commit?

matshenricson commented:
I just read up on what squashing means. Not sure what the idea is behind squashing...

RCasatta commented:
squashing help to keep the git history cleaner... like unifying 3 commits in one, but nevermind...

However Luca warned me there are a lot of warning on javadoc introduced by this PR that I missed, we can't have Warning from the command `mvn javadoc:javadoc` ([ output.txt](https://github.com/opentimestamps/java-opentimestamps/files/2992550/output.txt)) otherwise we cannot publish the lib on maven

matshenricson commented:
OK, I will have a look at these warnings.

matshenricson commented:
OK, I added the requested JavaDoc.

I think the code is worse now than before, because I have now added a lot of comments that really serves no purpose at all.

matshenricson commented:
Is it really the case that ONE SINGLE JavaDoc warning means that we can't publish from Maven?

RCasatta commented:
As far as I know, yes, we need to be warning free from javadoc to publish.

lvaccaro commented:
Usually squashing is used to merge together more commits in only one. I suggest to 
`git rebase -i master`, and keep your 1st commit as the destination and marks manually all following commits as "squash" (each commit marked as squash will be merged in the previous one, above in the list). Alternatively, you can also use `git merge --squash doc-changes` to squash all commits inside your branch. Remember to push with the flag `--force-with-lease` in order to update your remote repository.

I agree with @matshenricson, sometimes javadoc is still too much verbose, on the other hand, we has very few warnings right now. If you have problem running javadoc (maybe different version has different warnings), I can try to add the remaining comments and I ask you a review.

RCasatta commented:
merged after adding fix for latest javadoc warnings https://github.com/opentimestamps/java-opentimestamps/commit/1934d77fcdd9ddbbe3a0483d2935bfa50607833b
